tests/ui/lint/dangling-ptr/instantly-dangling-pointer-issue123613.rs

```diff
@@ -0,0 +1,13 @@
+#![deny(instantly_dangling_pointer)]
+
+const MAX_PATH: usize = 260;
+fn main() {
+    let str1 = String::with_capacity(MAX_PATH).as_mut_ptr();
```

Comment:
I wonder if it would make sense to suggest introducing a second binding in those simple `let` cases, aka.
```rust
    let str1 = String::with_capacity(MAX_PATH);
    let str1 = str1.as_mut_ptr();
```

Reply:
Probably yes, but I would prefer to do this as a separate follow-up PR if possible

Reply:
> I wonder if it would make sense to suggest introducing a second binding in those simple `let` cases, aka.
> 
> ```rust
>     let str1 = String::with_capacity(MAX_PATH);
>     let str1 = str1.as_mut_ptr();
> ```

After looking at the "regressed" code from crater, I can say that a lot of the time this suggestion will not fix the problem, and will just result in code like
```rust
let vec = some_iter.collect();
let ptr = vec.as_ptr();
return ptr;
```
or similar where the pointer still dangles eventually, even though we binded the temporary for now.

Sometimes it would make sense to suggest things like `.into_raw_parts()` instead of `.as_ptr()`, but then the user has to remember to do `::from_raw_parts()` if they want to avoid leaking...

Most of the places where this lint fires are not trivial to fix and actually require the person writing the code to use their brains and be more knowledgeable about what they are doing, which can only be achieved through more experience of writing `unsafe`, as well as some punches from the compiler, miri, or sanitizers.

I think that not providing this suggestion in places where it would help is less bad than improperly providing it in places where it wouldn't help.

Reply:
I still feel like we should point users towards doing something, even if it's not a general solution, and not a a structured suggestion.

It's really a shame to tell users something but not how to fix it.

Opened #132283 to track this.

Reply:
> It's really a shame to tell users something but not how to fix it.

It's even worse to suggest something that is *no a fix*. So if you are *sure* you have a correct suggestion, then sure, please make it -- but never do this unless you are sure. People will trust the suggestion and then be worse off than before.